### PR TITLE
MIR: Do not require END_BLOCK to always exist

### PR DESCRIPTION
**Pull request by nagisa**
Basically, all this does, is removing restriction for END_BLOCK to exist past the first invocation of RemoveDeadBlocks pass. This way for functions whose CFG does not reach the `END_BLOCK` end up not containing the block.

As far as the implementation goes, I’m not entirely satisfied with the `BasicBlock::end_block`. I had hoped to make `new` a `const fn` and then just have a `const END_BLOCK` private to mir::build, but it turns out that constant functions don’t yet support conditionals nor a way to assert.

### CODE REVIEW

File: src/librustc/mir/repr.rs [hunk 1]
```diff
@@ -216,6 +213,13 @@ impl BasicBlock {
         BasicBlock(index as u32)
     }
 
+    /// Returns a BasicBlock with index 1. This is actual end block (containing
+    /// the Return terminator) only during the building of MIR and should not be
+    /// used outside that.
+    pub const fn end_block() -> BasicBlock {
+        BasicBlock(1)
+    }
+
```
Comment (nikomatsakis) [inline src/librustc/mir/repr.rs]: Why have this at all? Why not make it somehow local to the `build` package? Ideally, it'd just be a field of the `Builder`.

Comment (nikomatsakis) [inline src/librustc/mir/repr.rs]: (Basically, I think we should go all the way and construct the END_BLOCK lazilly.)

**Comment by nagisa** [comment]
Updated

**Comment by nikomatsakis** [comment]
So I guess that this code will still build the return block, even for a function like `fn foo() -> ! { panic!() }`. I had sort of hoped to avoid that, but thinking a bit harder it seems sort of unavoidable. In any case I still like the lazy generation, since now it return/resume work analogously, which feels right.

**Pull request merged**